pkg/geo/geomfn/generate_points.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geomfn
+
+import (
+	"math"
+
+	"math/rand"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/cockroach/pkg/geo/geopb"
+	"github.com/cockroachdb/cockroach/pkg/geo/geos"
+	"github.com/cockroachdb/errors"
+	"github.com/twpayne/go-geom"
+)
+
+// GeneratePoints generates provided number of pseudo-random points for the input area.
+func GeneratePoints(g geo.Geometry, npoints int, seed int64) (geo.Geometry, error) {
+	if npoints < 0 {
+		return geo.Geometry{}, nil
+	}
+	if seed < 1 {
+		return geo.Geometry{}, errors.New("seed must be greater than zero")
+	}
+	pointsGeometry, err := geometryToPoints(g, npoints, seed)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("generating random points error: %v", err)
+	}
+	return pointsGeometry, nil
+}
+
+// geometryToPoints returns a MultiPoint geometry consisting of randomly generated points
+// that are covered by the geometry provided.
+// npoints is the number of points to return.
+// seed is the seed used by the random numbers generator.
+func geometryToPoints(g geo.Geometry, npoints int, seed int64) (geo.Geometry, error) {
+	var generatePointsFunction func(g geo.Geometry, npoints int, seed int64) (*geom.MultiPoint, error)
+	switch g.ShapeType() {
+	case geopb.ShapeType_Polygon:
+		generatePointsFunction = polygonGeometryToPoints
+	case geopb.ShapeType_MultiPolygon:
+		generatePointsFunction = multipolygonGeometryToPoints
+	default:
+		return geo.Geometry{}, errors.New("unsupported type")
+	}
+	// This is to be checked once we know Geometry type is supported,
+	// so that we can keep consistency with PostGIS implementation.
+	if npoints == 0 {
+		return geo.Geometry{}, nil
+	}
+	empty, err := IsEmpty(g)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not check if Geometry is empty: %v", err)
+	}
+	if empty {
+		return geo.Geometry{}, nil
+	}
+	mpt, err := generatePointsFunction(g, npoints, seed)
+	if err != nil {
+		return geo.Geometry{}, err
+	}
+	srid := g.SRID()
+	mpt.SetSRID(int(srid))
+	out, err := geo.MakeGeometryFromGeomT(mpt)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not transform geom.T into Geometry: %v", err)
+	}
+	return out, nil
+}
+
+func polygonGeometryToPoints(g geo.Geometry, npoints int, seed int64) (*geom.MultiPoint, error) {
+	area, err := Area(g)
+	if err != nil {
+		return nil, errors.Newf("could not calculate Polygon area: %v", err)
+	}
+	bbox := g.CartesianBoundingBox()
+	bboxWidth := bbox.HiX - bbox.LoX
+	bboxHeight := bbox.HiY - bbox.LoY
+	bboxArea := bboxHeight * bboxWidth
+	if area == 0.0 || bboxArea == 0.0 {
```

Comment:
That's how it works in PostGIS, so I wanted to double-check - Can the `bboxArea` be zero when `area` is not zero?

Reply:
All right, I removed `bboxArea` check here.